File: repo-templates/java-lambda-webapp/skeleton/src/main/java/com/example/orders/Handler.java
package com.example.orders;

import java.io.IOException;
import java.io.InputStream;
import java.io.UncheckedIOException;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.Map;
import java.util.Properties;
import java.util.UUID;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.amazonaws.services.lambda.runtime.events.APIGatewayV2HTTPEvent;
import com.amazonaws.services.lambda.runtime.events.APIGatewayV2HTTPResponse;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;

import software.amazon.awssdk.http.urlconnection.UrlConnectionHttpClient;
import software.amazon.awssdk.services.dynamodb.DynamoDbClient;

/**
 * The order API, answering API Gateway v2 HTTP requests. CloudFront forwards
 * /api/* here verbatim, so the paths handleRequest matches are the paths the
 * browser asks for.
 *
 * <p>ORDERS_TABLE is set by Wayfinder.yaml from the `db` component's output.
 * The image's CMD names {@code com.example.orders.Handler::handleRequest}.
 */
public final class Handler implements RequestHandler<APIGatewayV2HTTPEvent, APIGatewayV2HTTPResponse> {

    static final int DEFAULT_LIST_LIMIT = 50;
    static final int MAX_LIST_LIMIT = 200;

    static final ObjectMapper JSON = new ObjectMapper()
            .registerModule(new JavaTimeModule())
            .setPropertyNamingStrategy(PropertyNamingStrategies.SNAKE_CASE)
            .disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS)
            .disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);

    private final OrderStore store;
    private final String release;

    /** Called by the Lambda runtime, once per cold start. */
    public Handler() {
        String table = System.getenv("ORDERS_TABLE");
        if (table == null || table.isEmpty()) {
            throw new IllegalStateException("ORDERS_TABLE is not set; Wayfinder.yaml sets it from the db component");
        }
        DynamoDbClient client = DynamoDbClient.builder()
                .httpClientBuilder(UrlConnectionHttpClient.builder())
                .build();
        this.store = new DynamoOrderStore(client, table);
        this.release = readRelease();
        log("starting", Map.of("release", release, "table", table));
    }

    Handler(OrderStore store, String release) {
        this.store = store;
        this.release = release;
    }

    /**
     * Routes one request. An unknown path is a 404 rather than a fallthrough to
     * the SPA, so a mistyped API call does not come back as HTML the browser
     * then fails to parse as JSON.
     */
    @Override
    public APIGatewayV2HTTPResponse handleRequest(APIGatewayV2HTTPEvent req, Context context) {
        String path = req.getRawPath();
        String method = req.getRequestContext() == null || req.getRequestContext().getHttp() == null
                ? ""
                : req.getRequestContext().getHttp().getMethod();
        switch (path == null ? "" : path) {
            case "/api/healthz":
                return json(200, Map.of("status", "ok", "release", release));
            case "/api/orders":
                if ("POST".equals(method)) {
                    return placeOrder(req);
                }
                if ("GET".equals(method)) {
                    return listOrders(req);
                }
                return json(405, error(method + " is not allowed on /api/orders"));
            default:
                return json(404, error("no such endpoint: " + path));
        }
    }

    private APIGatewayV2HTTPResponse placeOrder(APIGatewayV2HTTPEvent req) {
        Order submitted;
        try {
            submitted = JSON.readValue(req.getBody() == null ? "" : req.getBody(), Order.class);
        } catch (JsonProcessingException e) {
            return json(400, error("the request body is not valid JSON: " + e.getOriginalMessage()));
        }
        if (submitted == null) {
            return json(400, error("the request body is not valid JSON: it is empty"));
        }
        String problem = submitted.problem();
        if (problem != null) {
            return json(400, error(problem));
        }

        Order order = submitted.accepted(UUID.randomUUID().toString(), Instant.now().truncatedTo(ChronoUnit.MICROS));
        try {
            store.put(order);
        } catch (RuntimeException e) {
            log("writing the order", Map.of("order_id", order.orderId(), "error", String.valueOf(e.getMessage())));
            return json(500, error("the order could not be saved"));
        }
        return json(201, order);
    }

    private APIGatewayV2HTTPResponse listOrders(APIGatewayV2HTTPEvent req) {
        int limit = DEFAULT_LIST_LIMIT;
        String raw = req.getQueryStringParameters() == null ? null : req.getQueryStringParameters().get("limit");
        if (raw != null && !raw.isEmpty()) {
            int n;
            try {
                n = Integer.parseInt(raw);
            } catch (NumberFormatException e) {
                n = 0;
            }
            if (n < 1) {
                return json(400, error("limit must be a positive whole number, got \"" + raw + "\""));
            }
            limit = Math.min(n, MAX_LIST_LIMIT);
        }

        List<Order> found;
        try {
            found = store.list(limit);
        } catch (RuntimeException e) {
            log("listing orders", Map.of("error", String.valueOf(e.getMessage())));
            return json(500, error("the orders could not be read"));
        }
        return json(200, Map.of("orders", found));
    }

    private static Map<String, String> error(String message) {
        return Map.of("error", message);
    }

    private static APIGatewayV2HTTPResponse json(int status, Object body) {
        String encoded;
        try {
            encoded = JSON.writeValueAsString(body);
        } catch (JsonProcessingException e) {
            status = 500;
            encoded = "{\"error\":\"the response could not be encoded\"}";
        }
        return APIGatewayV2HTTPResponse.builder()
                .withStatusCode(status)
                .withHeaders(Map.of("Content-Type", "application/json"))
                .withBody(encoded)
                .build();
    }

    /** One JSON line per event, which is what CloudWatch Logs Insights parses. */
    private static void log(String msg, Map<String, String> fields) {
        try {
            Map<String, Object> line = new java.util.LinkedHashMap<>();
            line.put("msg", msg);
            line.putAll(fields);
            System.out.println(JSON.writeValueAsString(line));
        } catch (JsonProcessingException e) {
            System.out.println(msg);
        }
    }

    private static String readRelease() {
        try (InputStream in = Handler.class.getResourceAsStream("/release.properties")) {
            if (in == null) {
                return "dev";
            }
            Properties p = new Properties();
            p.load(in);
            return p.getProperty("release", "dev");
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }
}

File: repo-templates/go-lambda-webapp/skeleton/internal/api/handler.go
// Package api serves the order API behind API Gateway. CloudFront forwards
// /api/* here verbatim, so the paths below are the paths the browser asks for.
package api

import (
	"context"
	"encoding/json"
	"log/slog"
	"net/http"
	"strconv"
	"time"

	"github.com/aws/aws-lambda-go/events"
	"github.com/google/uuid"

	"github.com/${{ .Repo.Organization }}/${{ .Repo.Name }}/internal/orders"
)

// defaultListLimit and maxListLimit bound ?limit= on GET /api/orders.
const (
	defaultListLimit = 50
	maxListLimit     = 200
)

// Handler answers API Gateway v2 HTTP requests.
type Handler struct {
	Store   orders.Store
	Release string
	Log     *slog.Logger
}

// Handle routes one request. An unknown path is a 404 rather than a fallthrough
// to the SPA, so a mistyped API call does not come back as HTML the browser
// then fails to parse as JSON.
func (h *Handler) Handle(ctx context.Context, req events.APIGatewayV2HTTPRequest) (events.APIGatewayV2HTTPResponse, error) {
	switch req.RawPath {
	case "/api/healthz":
		return h.json(http.StatusOK, map[string]string{"status": "ok", "release": h.Release})
	case "/api/orders":
		switch req.RequestContext.HTTP.Method {
		case http.MethodPost:
			return h.placeOrder(ctx, req)
		case http.MethodGet:
			return h.listOrders(ctx, req)
		}
		return h.json(http.StatusMethodNotAllowed, errorBody(req.RequestContext.HTTP.Method+" is not allowed on /api/orders"))
	}
	return h.json(http.StatusNotFound, errorBody("no such endpoint: "+req.RawPath))
}

func (h *Handler) placeOrder(ctx context.Context, req events.APIGatewayV2HTTPRequest) (events.APIGatewayV2HTTPResponse, error) {
	var o orders.Order
	if err := json.Unmarshal([]byte(req.Body), &o); err != nil {
		return h.json(http.StatusBadRequest, errorBody("the request body is not valid JSON: "+err.Error()))
	}
	if err := o.Validate(); err != nil {
		return h.json(http.StatusBadRequest, errorBody(err.Error()))
	}

	// The id, the total and the timestamp are ours to set: a client that sends
	// its own gets them replaced.
	o.OrderID = uuid.NewString()
	o.ReceivedAt = time.Now().UTC()
	o.Totalise()

	if err := h.Store.Put(ctx, &o); err != nil {
		h.Log.Error("writing the order", "order_id", o.OrderID, "error", err)
		return h.json(http.StatusInternalServerError, errorBody("the order could not be saved"))
	}
	return h.json(http.StatusCreated, &o)
}

func (h *Handler) listOrders(ctx context.Context, req events.APIGatewayV2HTTPRequest) (events.APIGatewayV2HTTPResponse, error) {
	limit := defaultListLimit
	if raw := req.QueryStringParameters["limit"]; raw != "" {
		n, err := strconv.Atoi(raw)
		if err != nil || n < 1 {
			return h.json(http.StatusBadRequest, errorBody("limit must be a positive whole number, got "+strconv.Quote(raw)))
		}
		limit = min(n, maxListLimit)
	}

	found, err := h.Store.List(ctx, limit)
	if err != nil {
		h.Log.Error("listing orders", "error", err)
		return h.json(http.StatusInternalServerError, errorBody("the orders could not be read"))
	}
	if found == nil {
		found = []*orders.Order{}
	}
	return h.json(http.StatusOK, map[string]any{"orders": found})
}

func errorBody(message string) map[string]string {
	return map[string]string{"error": message}
}

func (h *Handler) json(status int, body any) (events.APIGatewayV2HTTPResponse, error) {
	encoded, err := json.Marshal(body)
	if err != nil {
		h.Log.Error("encoding the response", "error", err)
		return events.APIGatewayV2HTTPResponse{
			StatusCode: http.StatusInternalServerError,
			Headers:    map[string]string{"Content-Type": "application/json"},
			Body:       `{"error":"the response could not be encoded"}`,
		}, nil
	}
	return events.APIGatewayV2HTTPResponse{
		StatusCode: status,
		Headers:    map[string]string{"Content-Type": "application/json"},
		Body:       string(encoded),
	}, nil
}
